File: 13_ilp/problems/00_simd_multicore/simd_multicore.f90
! Fortran には C/C++ のベクトル型拡張 (vector_size) に相当する機能が無い.
! そこで, 内側の漸化式ループを !$omp simd で SIMD 化し,
! 外側の独立な要素のループを !$omp parallel do でマルチコア並列化して,
! SIMD + 命令レベル並列 + マルチコアをコンパイラに任せる.

module lin_rec_mod
contains
  ! t = a*t + b を n 回繰り返す
  real(8) function lin_rec(a, b, c, n) result(t)
    real(8), intent(in) :: a, b, c
    integer(8), intent(in) :: n
    integer(8) :: j
    t = c
    !$omp simd
    do j = 1, n
       t = a * t + b
    end do
  end function lin_rec
end module lin_rec_mod

program simd_multicore
  use lin_rec_mod
  use omp_lib
  implicit none
  integer(8) :: m, n, i, err
  real(8), allocatable :: x(:)
  real(8) :: t0, t1, dt, flops
  character(len=32) :: arg

  m = 8
  n = 1000_8 * 1000 * 1000
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read(arg, *) m
  end if
  if (command_argument_count() >= 2) then
     call get_command_argument(2, arg); read(arg, *) n
  end if
  allocate(x(m))
  print '(A,I0,A,I0)', "m = ", m, ", n = ", n

  t0 = omp_get_wtime()
  ! TODO: 下の do ループの直前に !$omp parallel do を1行追加し, 外側 (互いに独立な要素) のループをマルチコアで並列化せよ.
  do i = 1, m
     x(i) = lin_rec(0.99d0, dble(i), 1.0d0, n)
  end do
  ! TODO: 上の do ループに対応する !$omp end parallel do を書け.
  t1 = omp_get_wtime()
  dt = t1 - t0

  err = 0
  do i = 1, m
     if (abs(x(i) - 100.0d0 * dble(i)) > 1.0d-3) then
        print '(A,I3,A,F9.3)', "x(", i, ") = ", x(i)
        err = err + 1
     end if
  end do
  if (err == 0) print '(A)', "OK"

  flops = 2.0d0 * dble(m) * dble(n)
  print '(A,F7.3,A)', "elapsed    : ", dt, "  sec"
  print '(A,ES9.2)', "flops      : ", flops
  print '(F8.3,A)', flops / dt * 1d-9, " GFLOPS"
end program simd_multicore

File: 13_ilp/problems/00_simd_multicore/simd_multicore.cpp
#include <cstdio>
#include <cstdlib>
#include <cassert>
#include <cmath>
#include <omp.h>

/* ベクトル長 nl. 2 のべき乗 (8, 16, 32, ...) に変えて性能を比べてみよ */
enum { nl = 8 };

typedef double doublev __attribute__((vector_size(sizeof(double) * nl),
                                      aligned(sizeof(double))));

/* スカラ u を全要素 u のベクトルに */
doublev uniform(double u) {
  doublev v;
  for (long i = 0; i < nl; i++) {
    v[i] = u;
  }
  return v;
}

/* u, u+1, u+2, ... からなるベクトル */
doublev range(double u) {
  doublev v;
  for (long i = 0; i < nl; i++) {
    v[i] = u + i;
  }
  return v;
}

/* ベクトル型 v を配列 a の先頭 nl 要素に書き込む */
void storev(double * a, doublev v) {
  for (int i = 0; i < nl; i++) {
    a[i] = v[i];
  }
}

/* b, t をベクトル型にした lin_rec.
   nl 個の独立な漸化式を同時に進める */
doublev lin_rec(double a, doublev b, double c, long n) {
  doublev t = uniform(c);
  for (long j = 0; j < n; j++) {
    t = a * t + b;
  }
  return t;
}

int main(int argc, char ** argv) {
  long m     = (1 < argc ? atol(argv[1]) : 8);
  long n     = (2 < argc ? atol(argv[2]) : 1000 * 1000 * 1000);
  double * x = (double *)calloc(sizeof(double), m);
  assert(x);
  printf("m = %ld, n = %ld, nl = %d\n", m, n, nl);
  /* 計測開始 */
  double t0 = omp_get_wtime();
  /* 計算本体 (nl 回ずつまとめて SIMD 実行). */
  // TODO: 下の for ループの直前に #pragma omp parallel for を1行追加し, 外側のループ (互いに独立) をマルチコアで並列化せよ.
  for (long i = 0; i < m; i += nl) {
    storev(&x[i], lin_rec(0.99, range(i) + 1, 1.0, n));
  }
  /* 計測終了 */
  double t1 = omp_get_wtime();
  double dt = t1 - t0;          /* sec */

  /* 答え表示 (x[i] = 100 * (i + 1) くらいのはず) */
  long err = 0;
  for (long i = 0; i < m; i++) {
    if (fabs(x[i] - 100 * (i + 1)) > 1.0e-3) {
      printf("x[%3ld] = %9.3f\n", i, x[i]);
      err++;
    }
  }
  if (err == 0) {
    printf("OK\n");
  }
  double flops = 2. * (double)m * (double)n;
  printf("elapsed    : %7.3f  sec\n", dt);
  printf("flops      : %.2e\n", flops);
  printf("%.3f GFLOPS\n", flops / dt * 1e-9);
  return 0;
}
